### PR TITLE
Refactor ast

### PR DESCRIPTION
**Pull request by hlim1**
This PR holds 'rectify.py' to rectify OFP generated XML, and modifications in the 'translate.py' and 'pyTranslate.py' to accommodate the changes in the AST XML that was generated by 'rectify.py'.

### CODE REVIEW

**Comment by adarshp** [comment]
@hlim1 A common cause of Travis errors that cannot be replicated locally is that there is some code in the developer's local copy of the repo that has not been committed and pushed. Can you check if this is the case for you? Run `git status` to see if there are any uncommitted changes, and if so, commit and push them.

**Comment by hlim1** [comment]
@adarshp Yes, all changes been committed and up-to-date on Github. I believe it has something to do with the test_program_analysis.py. If I’m correct, I have to add the test_program_analysis.py to call the new program rectify.py, I’m having a bit of issue with getting it work together. Since I do not want to mess up the current version of test_program_analysis.py, I have not committed the modification as it is not really working. It would be grateful if I can get some help with it. As we’ve discussed in the other email, I can come to your office tomorrow at 2 PM. Thank you!

**Comment by hlim1** [comment]
@adarshp I'm not really sure whether the code coverage is a big issue for merging into the master or not, but if so, do you mind giving me a suggestion what can be done to satisfy the coverage level of the current branch with the master? Thanks!

**Comment by adarshp** [comment]
> @adarshp I'm not really sure whether the code coverage is a big issue for merging into the master or not, but if so, do you mind giving me a suggestion what can be done to satisfy the coverage level of the current branch with the master? Thanks!

The coverage isn't a deal breaker - 60-65% is pretty good for academic research code. However, if you do want to increase coverage, you can go through the Codecov report to see which lines have changed coverage.

On your local machine, if you run `make test`, you can also see the lines that are missing coverage, like below:

image

**Comment by skdebray** [comment]
@hlim1 Please make sure that the preprocessor is restored appropriately.  @adarshp --- FYI, @hlim1 removed the current code for preprocessor.py while debugging some of the issues he had run into.

File: delphi/translators/for2py/rectify.py [hunk 1]
```diff
@@ -0,0 +1,1982 @@
...
+        "component-shape-spec-list__begin",
+        "explicit-shape-spec-list__begin",
+        "explicit-shape-spec",
+        "component-attr-spec",
+        "component-attr-spec-list",
+        "end-type-stmt",
+        "derived-type-def",
+    ]
+
+    """
```
Comment (adarshp) [inline delphi/translators/for2py/rectify.py]: Can you move this docstring (and the other similar ones in the rest of the module) into the function? This way they will be automatically associated with the function and show up in the autogenerated Sphinx documentation.

See https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html#example-google for examples.

File: delphi/translators/for2py/rectify.py [hunk 2]
```diff
@@ -0,0 +1,1982 @@
...
+        "function",
+        "internal-subprogram",
+        "internal-subprogram-part",
+        "prefix",
+        "exit",
+    ]
+
+    LOOP_CHILD_TAGS = ["header", "body", "format"]
+
+    DECLARATION_CHILD_TAGS = [
```
Comment (adarshp) [inline delphi/translators/for2py/rectify.py]: @hlim1 Usually, the convention is to put global variables in all uppercase letters - are these meant to be global variables (i.e. are they never going to change during the execution of the program?) If so, can you move them outside the class definition (see pyTranslate.py for examples).

**Comment by skdebray** [comment]
@hlim1 Does the new preprocessor.py cause some tests to fail because of undeclared variables?  (I'm not sure whether that's still an issue or not.)  If that's the case, let's address that when we meet tomorrow afternoon.

**Comment by hlim1** [comment]
@skdebray It looks like the tests are now all passing with the restored preprocessor as well.  The error was occurring when the test is done with the crop_yield, but it looks like it was fixed (not the issue of preprocessor, but in the other areas of the program) while Adarsh and I were debugging to get the pytest to work and pass. Thank you!

@adarshp Thanks for the suggestion. I moved the docstrings into the functions and lower cased the tag lists as suggested. If the code coverage is not a deal breaker, then I would like you to consider for merging it. Currently, I’m working on the goto_elimination branch, so I believe the code coverage may be an issue in that branch as well, so I will handle there instead of holding back on this branch to be merged and done. Thank you!

**Pull request merged**